feat: injectTypes (#11551)
withastro/astro: .changeset/mean-horses-kiss.md

> ---

Reviewer: Would love to see an example of an integration using the hook in here

Author: Isn't it already the case with the snippet below?

Reviewer: I mean in the actual integration hook, the snippet doesn't show where that function comes from. You can amend the snippet below too, no need for a new example necessarily

'astro': minor
---

> Adds a new `injectTypes` utility to the Integration API and refactors how typegen works

Reviewer 2: ```suggestion
Adds a new `injectTypes()` utility to the Integration API and refactors how type generation works
```

> Adds a new `injectTypes` utility to the Integration API and refactors how typegen works
> 
> Integrations can now use a new `injectTypes` utility exposed on the `astro:config:done` hook. It allows to inject types easily in the user's project. `filename` must ends with `".d.ts"` and `content` must be valid TypeScript (it will be formatted).
> 
> Under the hood, it will create a file at `/.astro/integrations/<normalized_integration_name>/<normalized_filename>.d.ts` and create references to it. `injectTypes` returns a URL to the normalized path.

Reviewer 2: Suggesting a couple tweaks here. In general in docs we show a `()` after function names to distinguish them from non-function variables:

```suggestion
Adds a new `injectTypes()` utility to the Integration API and refactors how type generation works

Integrations can now use a new `injectTypes()` utility in the `astro:config:done` hook. This utility allows an integration to inject types into a user's project. `filename` must end with `".d.ts"` and `content` must be valid TypeScript (it will be formatted).

Under the hood, it will create a file at `/.astro/integrations/<normalized_integration_name>/<normalized_filename>.d.ts` and create references to it. `injectTypes()` returns a URL to the normalized path.
```

```js

> const path = injectTypes({

Reviewer: What's the point of it returning the path?

Author: See what we're doing in core for content sync (ie. overriding the dts file with content updates)? I can imagine some people wanting to do this in integrations. Since we normalize the filename, people can't easily construct the path themselves so we give it to them

> const path = injectTypes({
>   filename: "types.d.ts",
>   content: "declare module 'virtual:integration' {}"
> })
> console.log(path) // URL

Author: ```suggestion
// my-integration/index.js
export default {
  name: 'my-integration',
  'astro:config:done': ({ injectTypes }) => {
    injectTypes({
      filename: "types.d.ts",
      content: "declare module 'virtual:my-integration' {}"
    })
  }
};
```

```

> Codegen has been refactored. As a user, you just need to update `src/env.d.ts`:

Reviewer 3: I believe it's worth mentioning that the old values are still valid

Author: ```suggestion
Codegen has been refactored. Although `src/env.d.ts` will continue to work as is, we recommend you update it:
```

```diff
- /// <reference types="astro/client" />
+ /// <reference path="../.astro/types.d.ts" />
- /// <reference path="../.astro/env.d.ts" />
- /// <reference path="../.astro/actions.d.ts" />

> ```

Reviewer 4: Question: Would this be a breaking change for users who run `astro check`? I think some people run that before `astro build`, and that may break their build pipeline.

Author: I think it would not be breaking. To be clear:
- it's fine to keep `astro/client`, it would just be duplicated
- `../.astro/types.d.ts` is not breaking since it already exists for CC
- As for any other core reference, it may break but since those features are marked as experimental I think it's fine
- As for `astro:db`, it may break as well?

@Princesseuh do you know if astro check would complain?

Reviewer: `astro check` runs the sync step first, so if the syncing "fix" the types then it'll be fine

Author: Outdated declarations are currently not removed

Reviewer: Ah, but okay. TypeScript does not report an error for triple slashes references that point to non-existing files, so there shouldn't be an error.

Reviewer 4: Thanks for the explanation. Then it seems fine then. Since the db package is also marked a minor (for v0), it's already a breaking release so changing the types path should be fine.

---
'astro': minor
---

Adds a new `injectTypes()` utility to the Integration API and refactors how type generation works

> Integrations can now use a new `injectTypes()` utility in the `astro:config:done` hook. This utility allows an integration to inject types into a user's project. `filename` must end with `".d.ts"` and `content` must be valid TypeScript (it will be formatted).
> 
> Under the hood, it will create a file at `/.astro/integrations/<normalized_integration_name>/<normalized_filename>.d.ts` and create references to it. `injectTypes()` returns a URL to the normalized path.

Reviewer 5: ```suggestion
Use `injectTypes()` in the `astro:config:done` hook to inject types into your user's project by adding a new a `*.d.ts` file.

The `filename` property will be used to generate a file at `/.astro/integrations/<normalized_integration_name>/<normalized_filename>.d.ts` and must end with `".d.ts"`.

The `content` property will create the body of the file, and must be valid TypeScript.

Additionally, `injectTypes()` returns a URL to the normalized path so you can overwrite its content later on, or manipulate it in any way you want.
```
How about pulling this from docs?

---
'astro': minor
---

> Adds a new `injectTypes()` utility to the Integration API and refactors how type generation works

Reviewer 5: ```suggestion
Adds a new [`injectTypes()` utility](https://docs.astro.build/en/reference/integrations-reference/#injecttypes-options) to the Integration API and refactors how type generation works
```
Just adding a link where documentation will be!
